linera-base/src/http.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Types used when performing HTTP requests.
+
+use custom_debug_derive::Debug;
+use linera_witty::{WitLoad, WitStore, WitType};
+use serde::{Deserialize, Serialize};
+
+use crate::hex_debug;
+
+/// An HTTP request.
+#[derive(Clone, Debug, Eq, PartialEq, WitLoad, WitStore, WitType)]
+#[witty(name = "http-request")]
+pub struct Request {
+    /// The [`Method`] used for the HTTP request.
+    pub method: Method,
+
+    /// The URL this request is intended to.
+    pub url: String,
+
+    /// The headers that should be included in the request.
+    pub headers: Vec<Header>,
+
+    /// The body of the request.
+    #[debug(with = "hex_debug")]
+    pub body: Vec<u8>,
+}
+
+impl Request {
+    /// Creates an HTTP GET [`Request`] for a `url`.
+    pub fn get(url: impl Into<String>) -> Self {
+        Request {
+            method: Method::Get,
+            url: url.into(),
+            headers: vec![],
+            body: vec![],
+        }
+    }
+
+    /// Creates an HTTP POST [`Request`] for a `url` with a `payload` that's an arbitrary bytes.
+    pub fn post(url: impl Into<String>, payload: impl Into<Vec<u8>>) -> Self {
+        Request {
+            method: Method::Post,
+            url: url.into(),
+            headers: vec![],
+            body: payload.into(),
+        }
+    }
+
+    /// Creates an HTTP POST [`Request`] for a `url` with a body that's the `payload` serialized to
+    /// JSON.
+    pub fn post_json(
+        url: impl Into<String>,
+        payload: &impl Serialize,
+    ) -> Result<Self, serde_json::Error> {
+        Ok(Request {
+            method: Method::Post,
+            url: url.into(),
+            headers: vec![Header::new("Content-Type", b"application/json")],
+            body: serde_json::to_vec(payload)?,
+        })
+    }
+
+    /// Adds a header to this [`Request`].
+    pub fn with_header(mut self, name: impl Into<String>, value: impl Into<Vec<u8>>) -> Self {
+        self.headers.push(Header::new(name, value));
+        self
+    }
+}
+
+/// The method used in an HTTP request.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, WitLoad, WitStore, WitType)]
+#[witty(name = "http-method")]
+pub enum Method {
+    /// A GET request.
+    Get,
+
+    /// A POST request.
+    Post,
+
+    /// A PUT request.
```

Comment:
Should we support all these methods?

Reply:
@jvff To be clear, we're going to have to disable everything but POST and GET for oracle queries.

Reply:
I'm not sure that's a good idea. It shouldn't matter what the method is for the HTTP request, and limiting the supported methods will limit application developers. As an example, publishing a blob in Walrus requires the PUT method.